src/vm.rs

```diff
@@ -716,107 +717,170 @@ pub trait Vm {
+	let one_second = Duration::from_secs(1);
+	let now = Instant::now();
+	let start = unsafe { rdtsc() };
+	if start > 0 {
+		loop {
+			if now.elapsed() >= one_second {
+				break;
 			}
 		}
-		None => Err(()),
+		let end = unsafe { rdtsc() };
+		Ok(((end - start) / MHZ_TO_HZ).try_into().unwrap())
+	} else {
+		Err(())
 	}
 }
 
-#[cfg(not(target_os = "linux"))]
-fn get_cpu_frequency_from_os() -> std::result::Result<u32, ()> {
-	//Not implemented yet for other systems
-	Err(())
-}
-
 #[cfg(test)]
 mod tests {
 	#[cfg(target_os = "linux")]
 	use crate::linux::tests::has_vm_support;
 
 	use super::*;
 
-	#[cfg(target_os = "linux")]
+	// test is derived from
+	// https://github.com/gz/rust-cpuid/blob/master/examples/tsc_frequency.rs
+	#[test]
+	fn test_detect_freq_from_cpuid() {
+		let cpuid = CpuId::new();
+		let has_tsc = cpuid
+			.get_feature_info()
+			.map_or(false, |finfo| finfo.has_tsc());
+
+		let has_invariant_tsc = cpuid
+			.get_extended_function_info()
+			.map_or(false, |efinfo| efinfo.has_invariant_tsc());
+
+		let tsc_frequency_hz = cpuid.get_tsc_info().map(|tinfo| {
+			if tinfo.nominal_frequency() != 0 {
+				tinfo.tsc_frequency()
+			} else if tinfo.numerator() != 0 && tinfo.denominator() != 0 {
+				// Skylake and Kabylake don't report the crystal clock, approximate with base frequency:
+				cpuid
+					.get_processor_frequency_info()
+					.map(|pinfo| pinfo.processor_base_frequency() as u64 * MHZ_TO_HZ)
+					.map(|cpu_base_freq_hz| {
+						let crystal_hz = cpu_base_freq_hz * tinfo.denominator() as u64
+							/ tinfo.numerator() as u64;
+						crystal_hz * tinfo.numerator() as u64 / tinfo.denominator() as u64
+					})
+			} else {
+				None
+			}
+		});
+
+		if has_tsc {
+			// Try to figure out TSC frequency with CPUID
+			println!(
+				"TSC Frequency is: {} ({})",
+				match tsc_frequency_hz {
+					Some(x) => format!("{} Hz", x.unwrap_or(0)),
+					None => String::from("unknown"),
+				},
+				if has_invariant_tsc {
+					"invariant"
+				} else {
+					"TSC frequency varies with speed-stepping"
+				}
+			);
+
+			// Check if we run in a VM and the hypervisor can give us the TSC frequency
+			cpuid.get_hypervisor_info().map(|hv| {
+				hv.tsc_frequency().map(|tsc_khz| {
+					let virtual_tsc_frequency_hz = tsc_khz as u64 * KHZ_TO_HZ;
+					println!(
+						"Hypervisor reports TSC Frequency at: {} Hz",
+						virtual_tsc_frequency_hz
+					);
+				})
+			});
+
+			// Determine TSC frequency by measuring it (loop for a second, record ticks)
+			let one_second = Duration::from_secs(1);
+			let now = Instant::now();
+			let start = unsafe { rdtsc() };
+			if start > 0 {
+				loop {
+					if now.elapsed() >= one_second {
+						break;
+					}
+				}
+				let end = unsafe { rdtsc() };
+				println!(
+					"Empirical measurement of TSC frequency was: {} Hz",
+					(end - start)
+				);
+			} else {
+				println!("Don't have rdtsc on stable!");
+				assert!(true);
```

Comment:
Do you need these?

Reply:
No really, just to trigger an error. Old x86 do not deliver a stable tsc